### PR TITLE
LG-3209: ToastProvider/useToast no longer forces rerender every time the stack changes

### PR DESCRIPTION
**Pull request by bruugey**

### CODE REVIEW

File: packages/toast/src/ToastContext/ToastReducer/useToastReducer.ts [hunk 1]
```diff
@@ -78,57 +78,67 @@ export const useToastReducer = (initialValue?: ToastStack) => {
...
+  }, []);
+
+  return useMemo(() => {
+    return {
+      pushToast,
+      popToast,
+      updateToast,
+      getToast,
+      clearStack,
+      stack,
```
[the remark this answers is not in the document]
Comment (TheSonOfThomp) [reply packages/toast/src/ToastContext/ToastReducer/useToastReducer.ts:141]: Or, I guess it's fine if the methods inside are the same?

File: .changeset/strong-onions-grow.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@leafygreen-ui/toast': patch
+---
+
+Memoizes functions and stack inside of useReducer
```
Comment (TheSonOfThomp) [inline .changeset/strong-onions-grow.md:5]: Should we add a comment why, and what bug this fixes?

**Review by TheSonOfThomp — CHANGES_REQUESTED** [review]
We should add a test suite to verify these fixes

File: packages/toast/src/ToastContext/ToastProvider/ToastProvider.tsx [hunk 3]
```diff
@@ -15,13 +15,14 @@ export const ToastProvider = ({
 }: React.PropsWithChildren<ToastProviderProps>) => {
   const { stack, ...toastFns } = useToastReducer(initialValue);
 
+  const cachedGetStack = useCallback(() => stack, [stack]);
```
[the remark this answers is not in the document]
Comment (bruugey) [reply packages/toast/src/ToastContext/ToastProvider/ToastProvider.tsx:18]: 👍

File: packages/toast/src/ToastContext/ToastProvider/ToastProvider.tsx [hunk 4]
```diff
@@ -15,13 +15,14 @@ export const ToastProvider = ({
...
+
+  const value = useMemo(() => {
```
[the remark this answers is not in the document]
Comment (bruugey) [reply packages/toast/src/ToastContext/ToastProvider/ToastProvider.tsx:20]: 👍

**Review by TheSonOfThomp — APPROVED** [review]
🍞

**Pull request merged**